File: src/csharp/Foundgine.Core/Semantic/Security/Execution/SecurityExecutionContextProviderExtensions.cs
namespace Foundgine.Core.Semantic.Security.Execution;

/// <summary>
/// Shared null-check helper for <see cref="ISecurityExecutionContextProvider"/>
/// consumers. Centralizing the missing-context failure here means every transport
/// adapter (MCP, GraphQL, future REST/gRPC) reports the same failure the same way,
/// instead of each adapter writing and maintaining its own message.
/// </summary>
public static class SecurityExecutionContextProviderExtensions
{
    /// <summary>
    /// Returns the current <see cref="SecurityExecutionContext"/>, or throws
    /// <see cref="UnauthorizedAccessException"/> if the host has not supplied one.
    /// </summary>
    /// <param name="provider">The host-supplied security context source.</param>
    /// <param name="transportName">
    /// Short transport identifier used in the exception message, e.g. <c>"GraphQL"</c>
    /// or <c>"MCP"</c>.
    /// </param>
    /// <param name="operationDescription">
    /// Short description of the operation being attempted, e.g. <c>"execution"</c>,
    /// <c>"capability discovery"</c>, or <c>"mutation execution"</c>.
    /// </param>
    public static SecurityExecutionContext RequireSecurityExecutionContext(
        this ISecurityExecutionContextProvider provider,
        string transportName,
        string operationDescription)
    {
        ArgumentNullException.ThrowIfNull(provider);
        if (string.IsNullOrWhiteSpace(transportName))
            throw new ArgumentException("Value cannot be null or whitespace.", nameof(transportName));
        if (string.IsNullOrWhiteSpace(operationDescription))
            throw new ArgumentException("Value cannot be null or whitespace.", nameof(operationDescription));

        return provider.GetSecurityExecutionContext()
               ?? throw new UnauthorizedAccessException(
                   $"{transportName} {operationDescription} requires a host-supplied SecurityExecutionContext. " +
                   $"The {transportName} caller cannot supply identity, tenant, audience, or warrant context.");
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/security/execution/SecurityExecutionContextProviderExtensions.java
package com.foundgine.core.semantic.security.execution;

import java.util.Objects;

/** Shared fail-closed helper for transport adapters. */
public final class SecurityExecutionContextProviderExtensions {
	private SecurityExecutionContextProviderExtensions() {
	}

	/**
	 * Returns the current trusted context or throws when the host has not
	 * established one. This intentionally fails closed.
	 */
	public static SecurityExecutionContext requireSecurityExecutionContext(ISecurityExecutionContextProvider provider,
			String transportName, String operationDescription) {
		Objects.requireNonNull(provider, "provider");
		if (transportName == null || transportName.isBlank())
			throw new IllegalArgumentException("Value cannot be null or whitespace: transportName");
		if (operationDescription == null || operationDescription.isBlank())
			throw new IllegalArgumentException("Value cannot be null or whitespace: operationDescription");

		SecurityExecutionContext context = provider.getSecurityExecutionContext();
		if (context == null) {
			throw new SecurityException(transportName + " " + operationDescription
					+ " requires a host-supplied SecurityExecutionContext. " + "The " + transportName
					+ " caller cannot supply identity, tenant, audience, or warrant context.");
		}
		return context;
	}
}
